### PR TITLE
How to render HTML pages as an HTTP server response Using Node.js

### PR DESCRIPTION
**Pull request by mosesreigns**
### Article checklist
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you placed your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

File: articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
+When developing web applications, you might need to render the HTML components inside your server. This helps you create interactive pages on the client side once a request to access these pages is made. There are various ways of hosting your HTML pages (website), for example;
+
+- Rendering your client-side on its own using frameworks for such as React or,
+- Rend the pages directly from the server - when the browser tries to access a route specified in your server, the server will load these HTML pages upon users' requests.
```
Comment (lalith1403) [inline articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md:4]: Break this into 2-3 simpler sentences.

File: articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+### Setting up
+
+- [Download Node.js](https://nodejs.org/en/) and install it. Run `node -v` to test if the installation was successful.
+
+```bash
+$ node -v
+v12.18.3 ## installed node.js version
+```
+
+- Create a Node.js project/directory. Initialize the project within the directory you have created. Use `npm init -y` to automatically initialize this Node.js project or check this [guide](/engineering-education/beginner-guide-to-npm/) to learn more about using NPM.
```
Comment (lalith1403) [inline articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md:31]: Mention installation of npm in the setting up section.

File: articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+});
+
+app.get("/", (req, res) => {
+  res.send("<html> <head>server Response</head><body><h1> This page was render direcly from the server <p>Hello there welcome to my website</p></h1></body></html>");
+});
+```
+
+Restart the server and open route `http://localhost:3000/` on the browser.
+
+![](server-rendered-html-elements.jpg)
```
Comment (lalith1403) [inline articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md:75]: Make sure to add text related to the content of the image. Also, modify all the image URLs to the following.
```alt text```

File: articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+});
+
+app.get("/", (req, res) => {
+  res.sendFile(__dirname + "/index.html");
+});
+```
+
+![](html-page-as-server-response.jpg)
+
+However, this did not load the CSS styling. `app.css` is a static file. To [static server files](https://expressjs.com/en/starter/static-files.html) included in the `index.html` use `express.static` as shown in the example below;
```
Comment (lalith1403) [inline articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md:191]: I think the second sentence is incomplete. Check that. It should be, to load static server files?

File: articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+
+![](server-rendered-html-elements.jpg)
+
+`res.send` is sending individual bits of HTML data to the server, but if we want to send an entire web page such as an `index.html`, we have to use something different.
+
+### Render HTML web pages as server responses
+
+The above can be very tiresome and not the type of code you want to write inside your server. In a normal web page, HTML elements are written in a `.html` file. This way, you can write all your HTML elements, including your CSS styling, to layout these elements. This way, you separate the server files and the HTML elements, thus a clean code setup.
+
+To render an HTML file into the server using Express, we use `res.sendFile()`. This read and render the data that you include in your HTML files. This transfers the file over to the browser whenever it makes a [Get request](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#Request_methods) to the server. The server submits a response status with HTML rendered web content as the message body.
```
Comment (lalith1403) [inline articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md:83]:
```suggestion
To render an HTML file into the server using Express, we use `res.sendFile()`. This reads and renders the data included in one's HTML files. This transfers the file to the browser upon a [GET request](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#Request_methods) to the server. The server submits a response status with HTML rendered web content as the message body.
```

File: articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+    <input type="email" name="youremail" placeholder="Email Address" required>
+    <input type="text" name="yourname" placeholder="Name" required>
+    <input class="subscribe-button" type="submit" value="Subscribe">
+  </form>
+</div>
+</body>
+</html>
+```
+
+`app.css`
```
Comment (lalith1403) [inline articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md:141]:
```suggestion
**File name**: _app.css_
```

Comment (lalith1403) [reply articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md:141]: Make the exact change everywhere file names are mentioned.

File: articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+});
+```
+
+Save the file and open `http://localhost:3000/` in the browser, and the server will send a web page as expected.
+
+![](html-page-with-css-as-server-response.jpg)
+
+### Parsering form data to the server using the HTML forms
+
+The server is running. It is now returning HTML form as a response to the client (browser). Whenever you access this server's route, the `Get` request will be executed from the browser. However, what would happen if you fill this form with the data and press the button _subscribe_?
```
Comment (lalith1403) [inline articles/rendering-html-pages-as-a-http-server-response-using-node-js/index.md:215]: Maintain consistency while mentioning commands such as GET, POST, etc.

**Review by lalith1403 — CHANGES_REQUESTED** [review]
Welcome to the program. This is a great article on rendering HTML pages using Node.js. I would like to point out a couple of points. 
1. While writing technical articles, we tend to stick to deliver the message from a third person's point of view. Usage of I, you, etc should be avoided. 
2. Maintain consistency in the usage of technical terms. GET, POST, Node.js, Express.js. Avoid mixing it up with get, Get, etc.

The outputs are great! The code works as mentioned. Make these changes and we can move on to the final round fo review.

**Comment by lalith1403** [comment]
@mosesreigns We will need a headshot pic and a short bio for the author's page. Refer to the contributing guidelines for more information. You may also refer to the authors folder in the main repository for more information.

**Comment by lalith1403** [comment]
@mosesreigns Add a hero image from Unsplash.com. Let me know if you have any doubts. Before modifying the files, make sure to pull the changes made to avoid merge conflicts.

**Comment by mosesreigns** [comment]
I have updated the article @lalith1403

**Comment by lalith1403** [comment]
What about author pic and bio?

**Comment by mosesreigns** [comment]
Adde that too @lalith1403

**Comment by lalith1403** [comment]
> Welcome to the program. This is a great article on rendering HTML pages using Node.js. I would like to point out a couple of points.
> 
> 1. While writing technical articles, we tend to stick to deliver the message from a third person's point of view. Usage of I, you, etc. should be avoided.
> 2. Maintain consistency in the usage of technical terms. GET, POST, Node.js, Express.js. Avoid mixing it up with get, Get, etc.
> 
> The outputs are great! The code works as mentioned. Make these changes and we can move on to the final round fo review.

I still see a lot of _you_ mentioned in the article. A third-person narrative suits technical writing better.

**Comment by mosesreigns** [comment]
> I still see a lot of _you_ mentioned in the article. A third-person narrative suits technical writing better.

Worked on most of the "you", Check the update @lalith1403

**Comment by hectorkambow** [comment]
plagiairsm check done

**Comment by hectorkambow** [comment]
@mosesreigns  👋  Welcome and great job submitting you 1st article to the EngEd program. 

We're happy to have you join the community - please take keen notes on any changes made by your peer reviewer or myself to apply those to future articles. 🚀  

Be sure to go over our resources page if you haven't already for more information and helpful links. 

cc @lalith1403

**Comment by hectorkambow** [comment]
@lalith1403  Please see the changes I made to the front matter 👍  

main thing is the list - the front matter does not need to be in list form.

**Pull request merged**